### PR TITLE
absolutify --project path

### PR DESCRIPTION
**Pull request by fredrikekre**
Fixes:
```
julia> Base.active_project()
"/home/fredrik/.julia/dev/Pkg/Project.toml"

julia> cd("Pkg")

julia> Base.active_project()
"/home/fredrik/.julia/dev/Pkg/Pkg/Project.toml"
```
note `Pkg/Pkg` in the second path.

### CODE REVIEW

File: base/initdefs.jl [hunk 1]
```diff
@@ -136,10 +136,10 @@ function init_load_path()
     end
     project = (JLOptions().project != C_NULL ?
         unsafe_string(Base.JLOptions().project) :
```
Comment (vtjnash) [inline base/initdefs.jl]: This should probably be handled in init.c, for conisistency with all other options that define paths

Comment (fredrikekre) [reply base/initdefs.jl]: The `abspath` call? Like https://github.com/JuliaLang/julia/blob/0846eb56548a52f07998f5ee354562fbff85a8f7/src/init.c#L587-L594?

Comment (fredrikekre) [reply base/initdefs.jl]: We still need the `abspath` here for relative paths from `JULIA_PROJECT` though.

File: base/initdefs.jl [hunk 2]
```diff
@@ -136,10 +136,10 @@ function init_load_path()
...
-        get(ENV, "JULIA_PROJECT", nothing))
+        get(ENV, "JULIA_PROJECT", ""))
     HOME_PROJECT[] =
         project == "" ? nothing :
-        project == "@." ? current_project() : project
+        project == "@." ? current_project() : abspath(project)
```
Comment (StefanKarpinski) [inline base/initdefs.jl]: If we resolve this at startup time, you can't do `--project=.` and have it follow you around. I'm not sure if resolving it at startup or having it follow you around is better.

File: base/initdefs.jl [hunk 3]
```diff
@@ -136,10 +136,10 @@ function init_load_path()
     end
     project = (JLOptions().project != C_NULL ?
         unsafe_string(Base.JLOptions().project) :
-        get(ENV, "JULIA_PROJECT", nothing))
+        get(ENV, "JULIA_PROJECT", ""))
```
Comment (StefanKarpinski) [inline base/initdefs.jl]: This has a behavioral side effect: it makes setting `JULIA_PROJECT` to the empty string behave the same as having it unset, which doesn't seem ideal and is something that I've generally avoided.

**Comment by StefanKarpinski** [comment]
> ```
> julia> Base.active_project()
> "/home/fredrik/.julia/dev/Pkg/Project.toml"
> ```

How did you start Julia to get this?

File: base/initdefs.jl (hunk 2 above)
Comment (KristofferC) [reply base/initdefs.jl]: Following around to the exact `pwd` is something I think you rarely want. 

```
➜  Pkg git:(kc/free_repo) ✗ ~/julia/julia --project=.
(Pkg) pkg>

shell> cd src/
/Users/kristoffer/.julia/dev/Pkg/src

(src) pkg>
```

What you might want is to have the same behavior as we had before where you try to find the project based on your current path but we wanted to write that as `@..` or something?

Comment (KristofferC) [reply base/initdefs.jl]: Could just change this to `project === nothing ? nothing : abspath(project)` to fix Stefans comment?

Comment (StefanKarpinski) [reply base/initdefs.jl]: There's no way to do that from the command-line or environment, but you can accomplish it now by doing `Base.HOME_PROJECT[] = "@."` to circumvent the startup time expansion and delay it until load time.

**Comment by fredrikekre** [comment]
> How did you start Julia to get this?

`julia-master --project=Pkg`

> If we resolve this at startup time, you can't do --project=. and have it follow you around. I'm not sure if resolving it at startup or having it follow you around is better.

I agree with Kristoffer here. The reason I found this is that I made a doc-build project with needed deps, but Documenter `cd`s to the build directory, so when I tried to load packages from doctests I got an error, since the env had changed.

> There's no way to do that from the command-line or environment, but you can accomplish it now by doing Base.HOME_PROJECT[] = "@."

We can skip the `abspath` call in `init_load_path` and only keep the one for the command line flag in `init.c`. That way you can set `JULIA_PROJECT=@.` to get that behaviour.

**Comment by fredrikekre** [comment]
I changed this to only absolutify `--project` paths. That is now consistent with how we treat other command line arguments that accepts paths. I left `JULIA_PROJECT` alone; we don't call `abspath` for other environment variables, so that is consistent. Also, I think it makes sense to `abspath` `--project` but not `JULIA_PROJECT` since `--project` paths are much more local to the process that you start, environment variables should "work" wherever you start julia, and hence it is expected that those are set as absolute paths from the user. Thoughts?

**Comment by StefanKarpinski** [comment]
> We can skip the abspath call in init_load_path and only keep the one for the command line flag in init.c. That way you can set JULIA_PROJECT=@. to get that behaviour.

That's not what everyone thought was the better default when I was working on that.

**Comment by StefanKarpinski** [comment]
> Which you can still accomplish with JULIA_PROJECT=. if you want.

No, they'd both be resolved at startup time. You'd have to do `Base.HOME_PROJECT[] = "."`.

**Comment by fredrikekre** [comment]
> No, they'd both be resolved at startup time.

Not in the state the PR is now.

**Comment by StefanKarpinski** [comment]
I don't think that changing the way the command-line option and the environment variable behave is a great idea. That seems like it will be a huge source of confusion.

**Comment by KristofferC** [comment]
Bump, would be nice to have this bug fixed in 1.0.1. What can we do to make progress here?

**Comment by fredrikekre** [comment]
> What can we do to make progress here?

Merge? :trollface: 

So the problem I have with the current behaviour is that the environment I explicitly choose to start julia with is changed under my feet, by code that does not do any Pkg operations (only `cd`). I understand that I maybe can't control this if the code explicitly changes environment, but I feel like environment unaware code should not change which code I will load.

I guess the question is if we should do the same for `JULIA_PROJECT` (which was in here from the beginning), I think that might be good to add back maybe.

**Comment by KristofferC** [comment]
Yes, unless we are explicit we don't want to accidentally activate random environments by changing `pwd` (which is a security concern).

@StefanKarpinski, do you have any issue with the current implementation? Barring comments, I will merge this in a couple of days.

**Comment by StefanKarpinski** [comment]
That makes me much happier with this :)

File: src/init.c [hunk 4]
```diff
@@ -593,6 +593,8 @@ static void jl_resolve_sysimg_location(JL_IMAGE_SEARCH rel)
         jl_options.outputbc = abspath(jl_options.outputbc, 0);
     if (jl_options.machine_file)
         jl_options.machine_file = abspath(jl_options.machine_file, 0);
+    if (jl_options.project && strncmp(jl_options.project, "@.", strlen(jl_options.project)) != 0)
```
Comment (vtjnash) [inline src/init.c]: can use `strcmp` – this would just be a more verbose and slower way of implementing that

**Pull request merged**